File: demo/src/main/java/br/com/stickyindexsample/view/layout/FastScroller.java
package br.com.stickyindexsample.view.layout;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.animation.ObjectAnimator;
import android.content.Context;
import android.support.annotation.NonNull;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.TextView;

import br.com.stickyindex.layout.IndexLayoutManager;
import br.com.stickyindex.model.Subscriber;
import br.com.stickyindexsample.R;

public class FastScroller extends LinearLayout implements Subscriber {
    private static final int BUBBLE_ANIMATION_DURATION = 250;
    private static final int TRACK_SNAP_RANGE = 5;

    private RecyclerView recyclerView;
    private TextView bubble;
    private View handle;
    private int height;
    private ObjectAnimator bubbleAnimator = null;

    private IndexLayoutManager stickyIndex;


    // CONSTRUCTORS ________________________________________________________________________________
    public FastScroller(final Context context, final AttributeSet attrs, final int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        initialise(context);
    }

    public FastScroller(final Context context) {
        super(context);
        initialise(context);
    }

    public FastScroller(final Context context, final AttributeSet attrs) {
        super(context, attrs);
        initialise(context);
    }

    private void initialise(Context context) {
        setOrientation(HORIZONTAL);
        setClipChildren(false);

        LayoutInflater inflater = LayoutInflater.from(context);
        inflater.inflate(R.layout.fastscroller, this, true);

        bubble = (TextView) findViewById(R.id.fastscroller_bubble);
        handle = findViewById(R.id.fastscroller_handle);

        bubble.setVisibility(INVISIBLE);
    }

    // CALLBACKs ___________________________________________________________________________________
    @Override
    protected void onSizeChanged(int w, int h, int oldw, int oldh) {
        super.onSizeChanged(w, h, oldw, oldh);
        height = h;
    }

    @Override
    public boolean onTouchEvent(@NonNull MotionEvent event) {
        final int action = event.getAction();

        switch (action) {
            case MotionEvent.ACTION_DOWN:
                if (event.getX() < handle.getX()) {
                    return false;
                }

                if (bubbleAnimator != null) {
                    bubbleAnimator.cancel();
                }

                if (bubble.getVisibility() == INVISIBLE) {
                    showBubble();
                }

                handle.setSelected(true);

            case MotionEvent.ACTION_MOVE:
                final float y = event.getY();

                setBubbleAndHandlePosition(y);
                setRecyclerViewPosition(y);

                return true;

            case MotionEvent.ACTION_UP:

            case MotionEvent.ACTION_CANCEL:
                handle.setSelected(false);
                hideBubble();

                return true;
        }

        return super.onTouchEvent(event);
    }

    // GETTERS AND SETTERS _________________________________________________________________________
    public void setRecyclerView(RecyclerView recyclerView) {
        this.recyclerView = recyclerView;
        //this.recyclerView.setOnScrollListener(scrollListener);
    }

    private void setRecyclerViewPosition(float y) {
        if (recyclerView != null) {
            int itemCount = recyclerView.getAdapter().getItemCount();
            float proportion;

            if (handle.getY() == 0) {
                proportion = 0f;
            } else if (handle.getY() + handle.getHeight() >= height - TRACK_SNAP_RANGE) {
                proportion = 1f;
            } else {
                proportion = y / (float) height;
            }

            int targetPos = getValueInRange(0, itemCount - 1, (int) (proportion * (float) itemCount));

            //Log.d("AppLog", "targetPos:" + targetPos);

            if (stickyIndex != null) {
                stickyIndex.update(recyclerView, 0, 1);
            }

            ((LinearLayoutManager) recyclerView.getLayoutManager()).scrollToPositionWithOffset(targetPos, 0);
            String bubbleText = ((TextGetter) recyclerView.getAdapter()).getTextFromAdapter(targetPos);

            bubble.setText(bubbleText);
        }
    }

    private int getValueInRange(int min, int max, int value) {
        int minimum = Math.max(min, value);
        return Math.min(minimum, max);
    }

    public void setBubbleAndHandlePosition(float y) {
        int bubbleHeight = bubble.getHeight();
        int handleHeight = handle.getHeight();

        handle.setY(getValueInRange(0, height - handleHeight, (int) (y - handleHeight / 2)));
        bubble.setY(getValueInRange(0, height - bubbleHeight - handleHeight / 2, (int) (y - bubbleHeight)));
    }

    public int getFastScrollHeight() {
        return height;
    }

    public void setStickyIndex(IndexLayoutManager stickyIndex) {
        this.stickyIndex = stickyIndex;
    }

    // UI __________________________________________________________________________________________
    private void showBubble() {
        bubble.setVisibility(VISIBLE);

        if (bubbleAnimator != null)
            bubbleAnimator.cancel();

        bubbleAnimator = ObjectAnimator.ofFloat(bubble, "alpha", 0f, 1f).setDuration(BUBBLE_ANIMATION_DURATION);
        bubbleAnimator.start();
    }

    private void hideBubble() {
        if (bubbleAnimator != null)
            bubbleAnimator.cancel();

        bubbleAnimator = ObjectAnimator.ofFloat(bubble, "alpha", 1f, 0f).setDuration(BUBBLE_ANIMATION_DURATION);
        bubbleAnimator.addListener(new AnimatorListenerAdapter() {
            @Override
            public void onAnimationEnd(Animator animation) {
                super.onAnimationEnd(animation);
                bubble.setVisibility(INVISIBLE);
                bubbleAnimator = null;
            }

            @Override
            public void onAnimationCancel(Animator animation) {
                super.onAnimationCancel(animation);
                bubble.setVisibility(INVISIBLE);
                bubbleAnimator = null;
            }
        });

        bubbleAnimator.start();
    }

    // SUBSCRIBER INTERFACE ________________________________________________________________________
    @Override
    public void update(RecyclerView rv, float dx, float dy) {
        View firstVisibleView=recyclerView.getChildAt(0);

        int firstVisiblePosition=recyclerView.getChildPosition(firstVisibleView);
        int visibleRange=recyclerView.getChildCount();
        int lastVisiblePosition=firstVisiblePosition+visibleRange;
        int itemCount=recyclerView.getAdapter().getItemCount();
        int position;

        if(firstVisiblePosition==0) {
            position = 0;
        } else if(lastVisiblePosition==itemCount) {
            position = itemCount;
        } else {
            position = (int) (((float) firstVisiblePosition / (((float) itemCount - (float) visibleRange))) * (float) itemCount);
        }

        float proportion=(float)position/(float)itemCount;

        setBubbleAndHandlePosition(getFastScrollHeight() * proportion);
    }
}

File: library/src/main/java/br/com/stickyindex/view/FastScroller.kt
package br.com.stickyindex.view

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.animation.ObjectAnimator
import android.animation.ObjectAnimator.ofFloat
import android.annotation.SuppressLint
import android.content.Context
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.MotionEvent
import android.view.MotionEvent.*
import android.widget.LinearLayout
import br.com.stickyindex.R
import kotlinx.android.synthetic.main.fastscroller_view.view.*
import java.lang.Math.max
import java.lang.Math.min

/**
 * Custom layout that adds fast scroller capabilities to a {@link RecyclerView}
 */
class FastScroller @JvmOverloads constructor(
        context: Context,
        attrs: AttributeSet? = null,
        defStyleAttr: Int = 0
) : LinearLayout(context, attrs, defStyleAttr) {

    companion object {
        private const val BUBBLE_ANIMATION_DURATION = 250
        private const val TRACK_SNAP_RANGE = 5
    }

    private var scrollHeight: Int = 0
    private lateinit var bubbleAnimator: ObjectAnimator
    private lateinit var recyclerView: RecyclerView

    init {
        orientation = HORIZONTAL
        clipChildren = false
        LayoutInflater.from(context).inflate(R.layout.fastscroller_view, this, true)
        bubble.visibility = INVISIBLE
    }

    /**
     * {@inheritDoc}
     */
    override fun onSizeChanged(width: Int, height: Int, oldWidth: Int, oldHeight: Int) {
        super.onSizeChanged(width, height, oldWidth, oldHeight)
        this.scrollHeight = height
    }

    /**
     * {@inheritDoc}
     */
    @SuppressLint("ClickableViewAccessibility")
    override fun onTouchEvent(event: MotionEvent): Boolean {
        return when (event.action) {
            ACTION_DOWN -> handleActionDown(event)
            ACTION_MOVE -> handleActionMove(event)
            ACTION_UP, ACTION_CANCEL -> handleActionUpCancel()
            else -> super.onTouchEvent(event)
        }
    }

    private fun handleActionDown(event: MotionEvent): Boolean {
        if (event.x < handle.x) return false
        if (bubble.visibility == INVISIBLE) showBubble()
        handle.isSelected = true
        setBubbleAndHandlePosition(event.y)
        setRecyclerViewPosition(event.y)
        return true
    }

    private fun handleActionMove(event: MotionEvent): Boolean {
        setBubbleAndHandlePosition(event.y)
        setRecyclerViewPosition(event.y)
        return true
    }

    private fun handleActionUpCancel(): Boolean {
        handle.isSelected = false
        hideBubble()
        return true
    }

    /**
     * Binds the fast-scroller to the provided {@link RecyclerView}, so they scroll event are synchronized
     */
    fun bindRecyclerView(recyclerView: RecyclerView) {
        this.recyclerView = recyclerView
        recyclerView.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrolled(recyclerView: RecyclerView?, dx: Int, dy: Int) {
                refreshScrollPosition()
            }
        })
    }

    private fun setRecyclerViewPosition(y: Float) {
        val targetPos = computeTargetPosition(y)
        val layoutManager = recyclerView.layoutManager as LinearLayoutManager
        layoutManager.scrollToPositionWithOffset(targetPos, 0)
        bubble.text = (recyclerView.adapter as FastScrollerLabelPublisher).getLabel(targetPos)
    }

    private fun computeTargetPosition(y: Float): Int {
        val itemCount = recyclerView.adapter.itemCount
        return getValueInRange(0, itemCount - 1, (computeProportion(y) * itemCount.toFloat()).toInt())
    }

    private fun computeProportion(y: Float) = when {
        handle.y == 0f -> 0f
        handle.y + handle.height >= scrollHeight - TRACK_SNAP_RANGE -> 1f
        else -> y / scrollHeight.toFloat()
    }

    private fun getValueInRange(min: Int, max: Int, value: Int) = min(max(min, value), max)

    private fun setBubbleAndHandlePosition(y: Float) {
        val bubbleHeight = bubble.height
        val handleHeight = handle.height
        handle.y = getValueInRange(0, scrollHeight - handleHeight, (y - handleHeight / 2).toInt()).toFloat()
        bubble.y = getValueInRange(0, scrollHeight - bubbleHeight - handleHeight / 2, (y - bubbleHeight).toInt()).toFloat()
    }

    private fun showBubble() {
        bubble.visibility = VISIBLE
        bubbleAnimator = createBubbleAnimator(0f, 1f)
        bubbleAnimator.start()
    }

    private fun hideBubble() {
        bubbleAnimator = createBubbleAnimator(1f, 0f)
        bubbleAnimator.addListener(object : AnimatorListenerAdapter() {
            override fun onAnimationEnd(animation: Animator) {
                bubble.visibility = INVISIBLE
            }

            override fun onAnimationCancel(animation: Animator) {
                bubble.visibility = INVISIBLE
            }
        })
        bubbleAnimator.start()
    }

    private fun createBubbleAnimator(v1: Float, v2: Float) =
            ofFloat(bubble, "alpha", v1, v2)
                    .setDuration(BUBBLE_ANIMATION_DURATION.toLong())

    private fun refreshScrollPosition() {
        val firstVisibleView = recyclerView.getChildAt(0)
        val firstVisiblePosition = recyclerView.getChildAdapterPosition(firstVisibleView)
        val visibleRange = recyclerView.childCount
        val lastVisiblePosition = firstVisiblePosition + visibleRange
        val itemCount = recyclerView.adapter.itemCount
        val position = computePosition(firstVisiblePosition, lastVisiblePosition, itemCount, visibleRange)
        val proportion = position.toFloat() / itemCount.toFloat()
        setBubbleAndHandlePosition(scrollHeight * proportion)
    }

    private fun computePosition(
            firstVisiblePosition: Int,
            lastVisiblePosition: Int,
            itemCount: Int,
            visibleRange: Int
    ) = when {
        firstVisiblePosition == 0 -> 0
        lastVisiblePosition == itemCount -> itemCount
        else -> (firstVisiblePosition.toFloat() / (itemCount.toFloat() - visibleRange.toFloat()) * itemCount.toFloat()).toInt()
    }
}
